code/Calendar.php

```diff
@@ -859,10 +859,71 @@ public static function getTimezoneNameNoLocationSpecific($value, $width = 'long'
+    /**
+     * Returns the localized name of a timezone, location-specific.
+     *
+     * @param string|\DateTime|\DateTimeZone $value the php name of a timezone, or a \DateTime instance or a \DateTimeZone instance for which you want the localized timezone name
+     * @param string $locale The locale to use. If empty we'll use the default locale set with {@link \Punic\Data::setDefaultLocale()}.
+     *
+     * @return string returns an empty string if the timezone has not been found, the timezone name otherwise
+     *
+     * @see http://www.unicode.org/reports/tr35/tr35-dates.html#Time_Zone_Goals
+     */
+    public static function getTimezoneNameLocationSpecific($value, $locale = '')
     {
-        return '';
+        $result = '';
+        if (!empty($value)) {
+            if (is_string($value)) {
+                $phpNames = static::getTimezonesAliases($value);
+                $timezone = null;
+                foreach ($phpNames as $phpName) {
+                    try {
+                        $timezone = new \DateTimeZone($phpName);
+                        break;
+                    } catch (\Exception $x) {
+                    }
+                }
+                if (!$timezone) {
+                    return '';
+                }
+                $location = $timezone->getLocation();
+            } elseif (is_a($value, '\\DateTime')) {
+                $timezone = $value->getTimezone();
+                $location = self::getTimezoneLocationFromDatetime($value);
+                if (empty($kind)) {
+                    if (intval($value->format('I')) === 1) {
+                        $kind = 'daylight';
+                    } else {
+                        $kind = 'standard';
+                    }
+                }
+            } elseif (is_a($value, '\\DateTimeZone')) {
+                $timezone = $value;
+                $location = $timezone->getLocation();
+            } else {
+                throw new Exception\BadArgumentType($value, 'string, DateTime, or DateTimeZone');
+            }
+
+            $name = '';
+            if (isset($location['country_code']) && $location['country_code'] !== '??') {
+                $data = Data::getGeneric('primaryZones');
+                if (isset($data[$location['country_code']]) ||
+                    count(\DateTimeZone::listIdentifiers(\DateTimeZone::PER_COUNTRY, $location['country_code'])) === 1) {
+                    $name = Territory::getName($location['country_code'], $locale);
+                }
+            }
+
+            if (!isset($name[0]) && substr($timezone->getName(), 0, 7) !== 'Etc/GMT') {
+                $name = static::getTimezoneExemplarCity($value, false, $locale);
+            }
+
+            if (isset($name[0])) {
+                $data = Data::get('timeZoneNames', $locale);
+                $result = sprintf($data['regionFormat'], $name);
```

Comment:
@c960657 What about using `regionFormat-type-daylight`/`regionFormat-type-standard` if we receive a `DateTime` (so that we can know if we want a daylight/standard name?